### PR TITLE
list group issues when deleting records

### PR DESCRIPTION
**Pull request by caburj**
### [FIX] web: should update group count when deleting records

**Issue:**

- Open a list view and group the records. Notice the count per group.
- BUG: When deleting a record under a group, the count in the group
  is not updated.

**Solution:**

When calling `deleteRecords` in the `DynamicGroupList`, we call
`deleteRecords` on each group's `list`. This "list" can be
`DynamicGroupList` or `DynamicRecordsList`. Regardless which instance
calls the `deleteRecords` method, we should take the number of the
deleted records and subtract it from the group's current count.

If it turned out that the calling object is a `DynamicGroupList`,
we should make sure that it returns the ids of the deleted records
so that the parent group can update its count as well, that's why
we are collecting the deleted record ids.

It's not so easy to follow but we can roughly imagine the structure
this way:

DynamicGroupList
-> DynamicGroupList[]
   -> DynamicRecordsList[]

Deleting items from the inner dynamic records list should result to
update of the parent group list, and eventually, to the update of the
root group list.

**Illustrations:**

Before: https://youtu.be/jlrCqB_uq0M
After: https://youtu.be/elLxt6v-zPI

---

### [FIX] web: update records of the current page when deleting

**Issue:**

When at the first page of the list, and the user deleted all/or some
of the records, the next page is not automatically displayed.

**Solution:**

This is because the first page (offset=0) is exempted from offset
adjustment (_adjustOffset) which is responsible in reloading the
records in the list.

In this commit, we propose to reload the list if the call to
`_adjustOffset` didn't call `load` and there are actually deleted
records.

**Illustration:**

Before fix: https://youtu.be/zQzugbCMSq8
After fix: https://youtu.be/DlE_aPo4ZZk

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by caburj** [comment]
@robodoo rebase-ff

File: addons/web/static/src/views/relational_model.js [hunk 1]
```diff
@@ -2194,9 +2194,17 @@ export class DynamicGroupList extends DynamicList {
...
     async deleteRecords() {
+        const allResIds = [];
         for (const group of this.groups) {
-            group.list.deleteRecords();
-        }
+            const resIds = await group.list.deleteRecords();
+            // Update the count of each group.
+            group.count = group.count - resIds.length;
+            // Collect the deleted resIds from each group.
+            allResIds.push(...resIds);
```
Comment (aab-odoo) [inline addons/web/static/src/views/relational_model.js:2213]: how does it behave when grouped on a many2many field, if the same record is deleted from several groups?

Comment (caburj) [reply addons/web/static/src/views/relational_model.js:2213]: Apparently, when "group by m2m" was introduced, deleting is not supported.

The following code links to the logic that hides the "Delete" button from the action menu items when the records are grouped by m2m.

https://github.com/odoo/odoo/blob/73af1d04d9d49eace16e3f88c32ce5d457b44b83/addons/web/static/src/views/list/list_controller.js#L271

**Review by aab-odoo — APPROVED** [review]
robodoo rebase-ff r+

**Pull request closed without merging**